### PR TITLE
Networks: Adds `restricted.networks.access` setting

### PR DESCRIPTION
**Pull request by tomponline**
- [x] Docs.
- [x] Restrict API access to networks specified in `restricted.networks.access` project config if `restricted=true`.
- [x] Restrict usage of networks in instance NIC devices.
- [x] Tests.

### CODE REVIEW

File: doc/api-extensions.md [hunk 1]
```diff
@@ -2035,4 +2035,14 @@ This updates the storage bucket API to return initial admin credentials at bucke
...
+It reports the total number of effective CPUs.
+
+## `projects_networks_restricted_access`
+
+Adds the `restricted.networks.access` project configuration key to indicate (as a comma-delimited list) which networks can be accessed inside the project.
+If not specified then all networks are accessible (assuming it is also allowed by the `restricted.devices.nic` setting).
+This also introduces a change whereby network access is controlled by the project's `restricted.devices.nic` setting.
+If `restricted.devices.nic` is set to `managed` (the default if not specified) then only managed networks are accessible.
+If `restricted.devices.nic` is set to `allow` then all networks are accessible (dependent on the `restricted.networks.access` setting).
+If `restricted.devices.nic` is set to `block` then no networks are accessible.
```
Comment (ru-fu) [inline doc/api-extensions.md:2047]:
```suggestion
Adds the `restricted.networks.access` project configuration key to indicate (as a comma-delimited list) which networks can be accessed inside the project.
If not specified, all networks are accessible (assuming it is also allowed by the `restricted.devices.nic` setting).
This also introduces a change whereby network access is controlled by the project's `restricted.devices.nic` setting:

* If `restricted.devices.nic` is set to `managed` (the default if not specified), only managed networks are accessible.
* If `restricted.devices.nic` is set to `allow`, all networks are accessible (dependent on the `restricted.networks.access` setting).
* If `restricted.devices.nic` is set to `block`, no networks are accessible.
```

File: doc/projects.md [hunk 2]
```diff
@@ -63,6 +63,7 @@ Key                                  | Type      | Condition             | Defau
 `restricted.devices.usb`             | string    | -                     | `block`                   | Prevents use of devices of type `usb`
 `restricted.idmap.uid`               | string    | -                     | -                         | Specifies the allowed host UID ranges allowed in the instance `raw.idmap` setting.
 `restricted.idmap.gid`               | string    | -                     | -                         | Specifies the allowed host GID ranges allowed in the instance `raw.idmap` setting.
+`restricted.networks.access`         | string    | -                     | -                         | Comma-delimited list of network names that are allowed for use in this project. If not set all networks are accessible (depending on the `restricted.devices.nic` setting)
```
Comment (ru-fu) [inline doc/projects.md:66]:
```suggestion
`restricted.networks.access`         | string    | -                     | -                         | Comma-delimited list of network names that are allowed for use in this project. If not set, all networks are accessible (depending on the `restricted.devices.nic` setting).
```

Comment (tomponline) [reply doc/projects.md:66]: I've added the full stop for now anyway.

Comment (ru-fu) [reply doc/projects.md]: The general rule is to add full-stops for complete sentences but not sentence fragments, but to be consistent within a table.
So ideally, we would make sure that all rows are either only fragments or full sentences (which is why I often move a second sentence into parentheses when I clean up a table).

If a table is inconsistent (and some rows have two sentences, for example), the rule is to use a full stop for all.

And what does definitely NOT work is having two sentences, but leave out the full stop for the second. ;)

Comment (ru-fu) [reply doc/projects.md]: An alternative could be
```suggestion
`restricted.networks.access`         | string    | -                     | -                         | Comma-delimited list of network names that are allowed for use in this project (if not set, all networks are accessible, depending on the `restricted.devices.nic` setting)
```

File: doc/api-extensions.md [hunk 3]
```diff
@@ -2035,4 +2035,14 @@ This updates the storage bucket API to return initial admin credentials at bucke
...
\ No newline at end of file
+It reports the total number of effective CPUs.
+
+## `projects_networks_restricted_access`
+
+Adds the `restricted.networks.access` project configuration key to indicate (as a comma-delimited list) which networks can be accessed inside the project.
+If not specified, all networks are accessible (assuming it is also allowed by the `restricted.devices.nic` setting, described below).
+
+This also introduces a change whereby network access is controlled by the project's `restricted.devices.nic` setting:
+* If `restricted.devices.nic` is set to `managed` (the default if not specified), only managed networks are accessible.
```
Comment (ru-fu) [inline doc/api-extensions.md:2046]: You need a blank line here.

File: doc/projects.md (hunk 2 above)
Comment (tomponline) [reply doc/projects.md]: I dont mind either way so will leave as your original suggestion.

File: doc/api-extensions.md (hunk 3 above)
Comment (tomponline) [reply doc/api-extensions.md:2047]: Fixed

**Review by ru-fu — APPROVED** [review]
Doc update looks good. :)

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Pull request merged**